Marshaling support for ints, floats, strs, lists, dicts, tuples

jakearmendariz opened the pull request:
My work on #3458. Marshaling still needs support for tuples, dicts, sets, etc. But I wanted to post a PR for some feedback (and possibly an early merge).

### How It Works
When marshaled, each buffer is prefixed with a byte indicating the type of data. Then while decoding we can check the first byte to see which datatype the buffer holds.

Lists are more complicated because they have different types, and can be recursive (lists inside of lists). Each list has a prefix of its size, and then each element's data is stored directly after another 8 bytes indicating its size.
Resulting in `[LIST_BYTE, <bytes for size>, <size><element>,<size><element>...]`

Already there are some improvement areas. Most notably, each list/element size is stored as usize (u64) and that can probably be changed to a u32.

### Testing
I added a python file for testing. It passes for RustPython, marshaling/unmarshaling primitives and lists. 

However, CPython marshals their data differently. I am not sure if they are supposed to match, however, I thought it was best to optimize the code to fit the RustPython Datastructures and not worry about how CPython handles marshaling.

### Closing
Please let me know if there are any problems or more idiomatic ways of writing this. I've never written rust code outside of personal projects so I expect some issues.

Thanks!

jakearmendariz commented on vm/src/stdlib/marshal.rs:
```diff
@@ -30,25 +78,108 @@ mod decl {
 
+    /// Read the next 8 bytes of a slice, convert to usize
+    /// Has side effect of increasing position pointer
+    fn eat_usize(bytes: &[u8], position: &mut usize) -> usize {
```
Errors could be stemming from this function. Not all architectures have a usize of 8 bytes (could be 2, 4, 8). So instead of reading usize, I am going to switch marshaling to use u32 despite architecture, then convert back to usize. So the only possible error is an element or list being longer than 2^32 bytes, in which case we send an error that it can't be marshaled

jakearmendariz commented on vm/src/protocol/buffer.rs:
```diff
@@ -11,7 +11,7 @@ use crate::{
     types::{Constructor, Unconstructible},
     PyObject, PyObjectPayload, PyObjectRef, PyObjectView, PyObjectWrap, PyRef, PyResult,
-    TryFromBorrowedObject, TypeProtocol, VirtualMachine,
+    TryFromBorrowedObject, TypeProtocol, VirtualMachine, PyValue,
```
Cargo fmt issues stemming from this file

youknowone commented:
Hi, I am sorry for late review, but I think I will take a few more vacations so it might be delayed a few more days, I am sorry!

We don't need data compatibility with CPython, because `marshal` requires the serialization used only in the exact python version. For fmt issue, please run `cargo fmt --all` before commit. It works great for most of cases.

I have no idea for mac/ubuntu test failures for now.

jakearmendariz commented:
I added support for dicts and tuples as well. I'll stop adding features until after a review. 

Thanks and I hope to hear back about this PR soon!

youknowone reviewed (changes requested):
I am really sorry for late review. To make a little excuse, recently I was horribly occupied by my main job.

The core approach looks good. I have a few suggestions about Rust and the project conventions.
I think you brought `test_marshal.py` from CPython.  Could you revise your commit message to mark the CPython version you brougt it from? here is a nice example: https://github.com/RustPython/RustPython/pull/3474/commits

I left other suggestions on the codes.

qingshi163 commented on vm/src/protocol/buffer.rs:
```diff
@@ -63,6 +63,15 @@ impl PyBuffer {
             .then(|| unsafe { self.contiguous_mut_unchecked() })
     }
 
+    pub fn from_byte_vector(bytes: Vec<u8>, vm: &VirtualMachine) -> Self {
+        let bytes_len = bytes.len();
+        PyBuffer::new(
+            PyValue::into_object(VecBuffer::from(bytes), vm),
+            BufferDescriptor::simple(bytes_len, true),
+            &VEC_BUFFER_METHODS,
+        )
+    }
```
I think it could be simplify as:
```
VecBuffer::from(bytes).into_pybuffer(true)
```

jakearmendariz commented on vm/src/builtins/dict.rs:
```diff
@@ -69,6 +69,10 @@ impl PyDict {
         &self.entries
     }
 
+    pub(crate) fn from_entries(entries: DictContentType) -> Self {
+        Self { entries }
+    }
```
Still working on this, everything else is updated. Thanks for all the feedback btw!

youknowone replied:
please let me know if you meet any blocker

youknowone commented:
Most of project using git including us prefer patches(PR) never use merge but only rebases. Could you try rebase if possible? When `upstream` is the official repository, `git rebase upstream/main` will be helpful. Otherwise it will be squashed into single commit except for CPython-originated parts.

jakearmendariz commented:
I tried to rebase some of the earlier commits by squashing each of them into the first. However, it looks like my `git push --force-with-lease` brought in many other commits to this PR. I still need to fetch upstream into this PR as well, but I am struggling to do this without creating more merge commits/problems.

Any tips haha? Should I create an entirely new PR, not sure if it's too far gone.
Thanks again for the help and feedback!

youknowone commented:
The history looks twisted a bit, but I think running `git fetch upstream` before `git rebase upstream/main` again will fix the problem. Please let me know if you feel hard to fix the problems yourself.

You don't need to make another PR, because one PR is always bound to one branch. You can fix it by force push whenever you want.

fanninpm commented:
I just use `git push --force` when I rebase.

youknowone commented:
@jakearmendariz I rebased it and pushed it to your github branch

jakearmendariz commented:
Sorry I have been so slow at responding and making changes, took a lot of hard classes this quarter and midterms have been killer.

A couple of notes, `test_marshal.py` was a file I created for testing marshaling, I didn't take it from CPython. Let me know if I should remove it or make it more clear that the file wasn't copied from CPython.

Also, I couldn't refactor `from_entries` to use for marshaling. Not sure if it's a blocker for this PR, or something I can try to fix in a subsequent PR.

I think that addresses all of the PR comments, everything else should have been fixed.

Thanks again for the help!

fanninpm commented:
> `test_marshal.py` was a file I created for testing marshaling, I didn't take it from CPython. Let me know if I should remove it or make it more clear that the file wasn't copied from CPython.

There's no real need to worry about that. A similar situation exists for `test_dis.py`.

youknowone commented:
For `test_marshal.py`, could you put it in `extra_tests/snippets` rather than `Lib/test`?

fanninpm commented:
> For `test_marshal.py`, could you put it in `extra_tests/snippets` rather than `Lib/test`?

I thought about that, but the snippets are run by _both_ CPython and RustPython, and I'm guessing there's a chance that the snippet would fail when CPython runs it.

youknowone commented:
I don't think putting original tests to `Lib/test` is a preferrable way. Rather than that, we don't need to test specific serialization form of marshal - because it is implementation-specific. I suggest to test roundtrip in the test instead of byte-by-byte comparison

jakearmendariz commented:
> I suggest to test roundtrip in the test instead of byte-by-byte comparison

Updated! No byte-str comparison, passes test for RustPython and CPython

youknowone commented:
@jakearmendariz thank you for long time effort. there is a error for `test_exceptions.py` related to marshal. Could you also check this error please?

```
======================================================================
ERROR: testRaising (test.test_exceptions.ExceptionTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/RustPython/RustPython/vm/pylib-crate/Lib/test/test_exceptions.py", line 67, in testRaising
    unlink(TESTFN)
  File "/home/runner/work/RustPython/RustPython/vm/pylib-crate/Lib/test/test_exceptions.py", line 63, in testRaising
    pass
  File "/home/runner/work/RustPython/RustPython/vm/pylib-crate/Lib/test/test_exceptions.py", line 63, in testRaising
    pass
  File "/home/runner/work/RustPython/RustPython/vm/pylib-crate/Lib/test/test_exceptions.py", line 61, in testRaising
    marshal.loads(b'')
ValueError: EOF where object expected.
```

jakearmendariz commented:
Changed to send an EOF error when an empty byte string is passed in!

youknowone reviewed (approved):
great! thank you very much!
